core/trino-main/src/main/java/io/trino/sql/planner/AdaptivePlanner.java

```diff
@@ -0,0 +1,429 @@
+            for (SubPlan sourceSubPlan : sourceSubPlans) {
+                PlanNode sourceNode = context.rewrite(sourceSubPlan.getFragment().getRoot(), sourceSubPlan.getChildren());
+                sourceNodesBuilder.add(sourceNode);
+                fragmentIdentifierToId.put(new FragmentIdentifier(node.getId(), sourceNode.getId()), sourceSubPlan.getFragment().getId());
+            }
+
+            List<PlanNode> sourceNodes = sourceNodesBuilder.build();
+            List<List<Symbol>> inputs = sourceNodes.stream().map(PlanNode::getOutputSymbols).collect(toImmutableList());
+            PartitioningScheme partitioningScheme = node.getSourceFragmentIds().stream()
+                    .map(runtimeInfoProvider::getPlanFragment)
+                    .map(PlanFragment::getOutputPartitioningScheme)
+                    .findFirst()
+                    .orElseThrow();
+
+            return new ExchangeNode(
+                    node.getId(),
+                    node.getExchangeType(),
+                    REMOTE,
+                    partitioningScheme,
+                    sourceNodes,
+                    inputs,
+                    node.getOrderingScheme());
+        }
+
+        private Map<FragmentIdentifier, PlanFragmentId> getFragmentIdentifierToId()
+        {
+            return ImmutableMap.copyOf(fragmentIdentifierToId);
+        }
+    }
+
+    private static class CurrentPlanRewriter
+            extends SimplePlanRewriter<List<SubPlan>>
+    {
+        @Override
+        public PlanNode visitAdaptivePlanNode(AdaptivePlanNode node, RewriteContext<List<SubPlan>> context)
+        {
+            verify(
+                    !containsAdaptivePlanNode(node.getCurrentPlan()),
+                    "Adaptive plan node cannot have a nested adaptive plan node");
+            return node.getCurrentPlan();
+        }
+    }
+
+    private static class InitialPlanRewriter
+            extends SimplePlanRewriter<List<SubPlan>>
+    {
+        @Override
+        public PlanNode visitAdaptivePlanNode(AdaptivePlanNode node, RewriteContext<List<SubPlan>> context)
+        {
+            verify(
+                    !containsAdaptivePlanNode(node.getInitialPlan()),
+                    "Adaptive plan node cannot have a nested adaptive plan node");
+            return node.getInitialPlan();
+        }
+    }
+
+    private static boolean containsAdaptivePlanNode(PlanNode node)
+    {
+        return PlanNodeSearcher.searchFrom(node)
+                .whereIsInstanceOfAny(AdaptivePlanNode.class)
+                .matches();
+    }
+
+    private static Map<Symbol, Type> getFilteredSymbols(PlanNode node, TypeProvider types)
+    {
+        Set<Symbol> dependencies = SymbolsExtractor.extractOutputSymbols(node);
+        return Maps.filterKeys(types.allTypes(), in(dependencies));
+    }
+
+    public record FragmentIdentifier(PlanNodeId stageExchangeId, PlanNodeId stageRootId)
```

Comment:
This name clashes with `PlanFragmentId`. 
Maybe just call it `ExchangeSourceId` and name fields `exchangeId`, `sourceId`